### PR TITLE
DTO Validators Part 2

### LINKED ISSUE #497: Fluent validation issue
In case of an error, the services return an error code. But this is not a flexible and comfortable decision. It was decided to replace it with fluent validation. Create your own issue, specific to each service in the project and change the old error codes with fluent validation (example you can find in EventService).  Fluent validation should check only the main problems. For example, id cannot be null and should be more than zero. The fluent validator shouldn't make any calls to the database - if an entity doesn't exist the repository should just throw NotFoundException and then ExceptionHandleMiddleware must catch it. Also, add unit tests to your pull request to check if you didn't break some important validation logic.

### LINKED ISSUE #506: Add validators for the rest of the DTOs (FluentValidation)
### Issue
Currently (by the latest commit https://github.com/ita-social-projects/WhatBackend/commit/0f415f4a1b1622a2a11bc0879b5633980dec18a3) only some of the DTOs that come from API user have corresponding validator classes created (`AccountDtoValidator`, `ResetPasswordDtoValidator`, etc).

### Solution
What we need to do is to create a corresponding validator classes for the rest of DTOs that require validation, and use [FluentValidation](https://fluentvalidation.net/) instead of built-in validation attributes.

This way we are **uncoupling** validation rules from the models of the core application level.

---

Just to clarify even more. We need to replace these validation attributes like these:
https://github.com/ita-social-projects/WhatBackend/blob/0f415f4a1b1622a2a11bc0879b5633980dec18a3/CharlieBackend.Core/DTO/Lesson/CreateLessonDto.cs#L15

with validation classes like this:

https://github.com/ita-social-projects/WhatBackend/blob/0f415f4a1b1622a2a11bc0879b5633980dec18a3/CharlieBackend.Api/Validators/AccountDTOValidators/CreateAccountDtoValidator.cs#L7

### LINKED ISSUE #511: [Creating Schedule feature] Incorrect syntax processing for date pattern type 2 in the body for POST request /api/schedules
Environment: Postman v8.5.0
Reproducible: always.
Build found: 899f397 commit from https://github.com/ita-social-projects/WhatBackend/tree/dev

Preconditions

Postman is installed and launched.
At least one active Mentor, one theme, one group are existed.
Authorization Admin/Secretary Token is applied for the new request.

Steps to reproduce

Make POST request, URI: 'https://whatbackend.azurewebsites.net/api/schedules'. Body: 
{
  "pattern": {
    "type": 2,
    "interval": 1,
    "dates": [9]
  },
  "range": {
    "startDate": "2021-06-01T11:00:00",
    "finishDate": "2021-07-30T13:00:00"
  },
  "context": {
    "groupID": 1,
    "themeID": 1,
    "mentorID": 1
  }
}

**Actual result**

Code: 400, Body: 
{
    "errors": {
        "Pattern.Dates": [
            "The field Dates must be between 1 and 31."
        ]
    },
    "type": "https://tools.ietf.org/html/rfc7231#section-6.5.1",
    "title": "One or more validation errors occurred.",
    "status": 400,
    "traceId": "|23cf540-4ea91da23f832b24."
}

**Expected result**

Code: 200 and response body with appropriate storage value, IDs for events and ID for event occurrance is:
{
    "id": int,
    "studentGroupId": 1,
    "eventStart": "2021-06-01T11:00:00",
    "eventFinish": "2021-07-30T13:00:00",
    "pattern": 2,
    "events": [
        {
            "id": int,
            "eventOccuranceId": int,
            "studentGroupId": 1,
            "themeId": 1,
            "mentorId": 1,
            "lessonId": null,
            "eventStart": "2021-06-09T11:00:00",
            "eventFinish": "2021-06-09T13:00:00"
        },
        {
            "id": int,
            "eventOccuranceId": int,
            "studentGroupId": 1,
            "themeId": 1,
            "mentorId": 1,
            "lessonId": null,
            "eventStart": "2021-07-09T11:00:00",
            "eventFinish": "2021-07-09T13:00:00"
        }
    ],
    "storage": int
}

**Labels to be added**
"Bug", Priority ("High"), Severity ("High"), Type ("Functional").

### PR DESCRIPTION
**Pull request by DemetriosKoras**

### CODE REVIEW

File: CharlieBackend.Api/Controllers/MentorsController.cs [hunk 1]
```diff
@@ -41,6 +41,10 @@ public MentorsController(IMentorService mentorService, ILessonService lessonServ
         [HttpPost("{accountId}")]
         public async Task<ActionResult> PostMentor(long accountId)
         {
+            if (!ModelState.IsValid)
```
Comment (andrii-developer) [inline CharlieBackend.Api/Controllers/MentorsController.cs]: I don't think that this code checks something.

File: CharlieBackend.Api/Validators/HomeworkStudentDTOValidators/HomeworkStudentDtoValidator.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
+﻿using CharlieBackend.Business.Helpers;
+using CharlieBackend.Core.DTO.HomeworkStudent;
+using FluentValidation;
+
```
Comment (andrii-developer) [inline CharlieBackend.Api/Validators/HomeworkStudentDTOValidators/HomeworkStudentDtoValidator.cs]: Too many blank lines.

**Comment by andrii-developer** [comment]
Please remove redundant blank lines in all files. 
Let's walk through all validators in our meeting and discuss each of the rules.

**Comment by gingerbread-monster** [comment]
> Please remove redundant blank lines in all files.
> Let's walk through all validators in our meeting and discuss each of the rules.
https://github.com/ita-social-projects/WhatBackend/pull/519#issuecomment-867114297

Oof
You're gonna have so much fun 😄 
Good luck

File: CharlieBackend.Api/Controllers/MentorsController.cs (hunk 1 above)
Comment (gingerbread-monster) [inline CharlieBackend.Api/Controllers/MentorsController.cs]: https://github.com/ita-social-projects/WhatBackend/pull/519#discussion_r657413864
@andrii-developer is right. You don't need to manually check the ModelState in Web API controller.

>Web API controllers don't have to check `ModelState.IsValid` if they have the `[ApiController]` attribute. In that case, an automatic HTTP 400 response containing error details is returned when model state is invalid.
[Source: Model state section](https://docs.microsoft.com/en-us/aspnet/core/mvc/models/validation#model-state)

**Review by gingerbread-monster — COMMENTED** [review]
You could also consider creating a class hierarchy for your DTOs. This could help reduce the number of validation classes and get rid of repetitive validation code.

Let's say you created `BaseLessonDto` with some properties that are shared in every other specific Homework DTO. So your `CreateLessonDtoValidator` would look something like this:
```cs
public class CreateLessonDtoValidator : AbstractValidator<CreateLessonDto>
{
    public CreateLessonDtoValidator(IValidator<LessonBaseDto> baseValidator)
    {
        Include(baseValidator);

        // ...
        // some create-specific validation rules
    }
}
```

And your validator for base DTO would look something like this:
```cs
public class LessonBaseDtoValidator : AbstractValidator<LessonBaseDto>
{
    public LessonBaseDtoValidator()
    {
        // validation rules for common properties
    }
}
```

But it's up to you :)

---
See fluent validation docs for reference: [including rules](https://docs.fluentvalidation.net/en/latest/including-rules.html), or maybe [inheritance validation](https://docs.fluentvalidation.net/en/latest/inheritance.html)

**Review by Pasha1968 — APPROVED** [review]

**Review by andrii-developer — APPROVED** [review]

**Review by Vso1o — APPROVED** [review]

**Pull request merged**